### PR TITLE
[dua] fix potential issue in DUA.rsp & DUA.ntf processing

### PR DESCRIPTION
**Pull request by simonlingoogle**

### CODE REVIEW

**Comment by simonlingoogle** [comment]
@librasungirl @edmont any comments on this fix?

**Comment by edmont** [comment]
@simonlingoogle any background about the fixed issue?

**Comment by simonlingoogle** [comment]
> @simonlingoogle any background about the fixed issue?

There is a potential bug in `DuaManager::ProcessDuaResponse(Coap::Message &aMessage)`. 

The original code finds the Child by `Child *child = Get<ChildTable>().GetChildAtIndex(mChildIndexDuaRegistering);`, which assumes that the `DUA.rsp` must be targeting the current registering Child. This is true for `DUA.rsp`, but not correct for `DUA.ntf` because BBR may send `DUA.ntf` at any time to any device.

So, this comment changes `DuaManager::ProcessDuaResponse` code to search for the correct Child with the target DUA instead of using `mChildIndexDuaRegistering`. 

This commit also contains other refactoring on `mChildIndexDuaRegistering` usage.

**Review by edmont — APPROVED** [review]
Thanks, looks like a proper fix to me.

File: src/core/thread/dua_manager.cpp [hunk 1]
```diff
@@ -645,27 +649,41 @@ Error DuaManager::ProcessDuaResponse(Coap::Message &aMessage)
 #endif
 #if OPENTHREAD_FTD && OPENTHREAD_CONFIG_TMF_PROXY_DUA_ENABLE
     {
-        Child *child = Get<ChildTable>().GetChildAtIndex(mChildIndexDuaRegistering);
+        Child *  child = nullptr;
+        uint16_t childIndex;
+
+        for (Child &iter : Get<ChildTable>().Iterate(Child::kInStateValidOrRestoring))
```
Comment (librasungirl) [inline src/core/thread/dua_manager.cpp:655]: Only iterate valid child here? Note that the parent only register for valid child.
https://github.com/openthread/openthread/blob/3a953086730eaad12f4319f5608e42bbc333727d/src/core/thread/dua_manager.cpp#L479-L488

File: src/core/thread/dua_manager.cpp [hunk 2]
```diff
@@ -645,27 +649,41 @@ Error DuaManager::ProcessDuaResponse(Coap::Message &aMessage)
...
-        VerifyOrExit(child->HasIp6Address(target), error = kErrorNotFound);
+
+        childIndex = Get<ChildTable>().GetChildIndex(*child);
 
         switch (status)
         {
         case ThreadStatusTlv::kDuaSuccess:
             // Mark as Registered
-            mChildDuaRegisteredMask.Set(mChildIndexDuaRegistering, true);
+            if (mChildDuaMask.Get(childIndex))
```
Comment (librasungirl) [inline src/core/thread/dua_manager.cpp:672]: Is this ` if` condition necessary?

File: src/core/thread/dua_manager.cpp (hunk 1 above)
Comment (simonlingoogle) [reply src/core/thread/dua_manager.cpp:655]: resolved.

**Review by librasungirl — APPROVED** [review]

**Review by jwhui — APPROVED** [review]

**Pull request merged**